baas-node-state/src/main/scala/com/ing/baker/baas/state/ServiceDiscovery.scala

```diff
@@ -106,39 +94,62 @@ object ServiceDiscovery extends LazyLogging {
 
-    val stream = for {
-      cacheInteractions <- Stream.eval(Ref.of[IO, List[InteractionInstance]](List.empty))
-      cacheRecipeListeners <- Stream.eval(Ref.of[IO, Map[RecipeName, List[RecipeListener]]](Map.empty))
-      cacheBakerListeners <- Stream.eval(Ref.of[IO, List[BakerListener]](List.empty))
-      service = new ServiceDiscovery(cacheInteractions, cacheRecipeListeners, cacheBakerListeners)
-      updateServices = fetchServices(namespace, api)
-        .flatMap { currentServices =>
-          List(
-            buildInteractions(currentServices).flatMap(cacheInteractions.set),
-            cacheRecipeListeners.set(buildRecipeListeners(currentServices)),
-            cacheBakerListeners.set(buildBakerListeners(currentServices))
-          ).parSequence
+    def updateServicesWith(
+      currentServices: Ref[IO, List[Service]],
+      cacheInteractions: Ref[IO, List[InteractionInstance]],
+      cacheRecipeListeners: Ref[IO, Map[RecipeName, List[RecipeListener]]],
+      cacheBakerListeners: Ref[IO, List[BakerListener]]
+    )(event: K8SWatchEvent[Service]): IO[Unit] = {
+      for {
+        services <- event._type match {
+          case EventType.ADDED =>
+            currentServices.updateAndGet(event._object :: _)
+          case EventType.DELETED =>
+            currentServices.updateAndGet(_.filterNot(_ == event._object))
+          case EventType.MODIFIED =>
+            currentServices.updateAndGet(_.map(service => if (service.metadata.uid == event._object.metadata.uid) event._object else service))
+          case EventType.ERROR =>
+            IO(logger.error(s"Event type ERROR on service watch for service ${event._object}")) *> currentServices.get
         }
-      updater = Stream.fixedRate(5.seconds).evalMap(_ => updateServices)
-      _ <- Stream.eval(updateServices).concurrently(updater)
-    } yield service
+        _ <- List(
+          buildInteractions(services).flatMap(cacheInteractions.set),
+          cacheRecipeListeners.set(buildRecipeListeners(services)),
+          cacheBakerListeners.set(buildBakerListeners(services))
+        ).parSequence
+      } yield ()
+    }
+
+    val createServiceDiscovery = for {
+      currentServices <- Ref.of[IO, List[Service]](List.empty)
+      cacheInteractions <- Ref.of[IO, List[InteractionInstance]](List.empty)
+      cacheRecipeListeners <- Ref.of[IO, Map[RecipeName, List[RecipeListener]]](Map.empty)
+      cacheBakerListeners <- Ref.of[IO, List[BakerListener]](List.empty)
+      service = new ServiceDiscovery(cacheInteractions, cacheRecipeListeners, cacheBakerListeners)
+      updateServices = updateServicesWith(currentServices, cacheInteractions, cacheRecipeListeners, cacheBakerListeners) _
+      killSwitch <- IO {
+        k8s.watchAllContinuously[Service]()
+          .viaMat(KillSwitches.single)(Keep.right)
+          .toMat(Sink.foreach(updateServices(_).unsafeRunAsyncAndForget()))(Keep.left)
```

Comment:
what are the implications of running async `updateServices()`? 
is there a guarantee that there won't ever be concurrent updates?

Reply:
This is the required interoperability between akka streams and cats effect, Sink.foreach is meant to fire side effects per produced value... side effects which are contained in the IO program. This would not be required if we had a single library that would continuously contain side effects ofc, but skuber imposes akka streams here

Reply:
I think the comment was that it's better to use `mapAsync` and inside use  `.unsafeRunAsync` that will run the stream with waiting the results. If `updateServices` is ok to be run concurrently, it's better to use parralel features on stream, but still wait the result of `updateServices`. That will still allow throttling and backpreassure features enabled.

Reply:
I believe I understand both of your messages better now, so first:
q: is there a guarantee that there won't ever be concurrent updates?
a: yes, we are using cats.effect.concurrent.Ref which documentation reads:

 * An asynchronous, concurrent mutable reference.
 *
 * Provides safe concurrent access and modification of its content

q: it's better to use mapAsync and inside use .unsafeRunAsync that will run the stream with waiting the results
a: I agree, I have changed it to a Sink.foreachAsync and transformed the IO of the update to a future to make the akka stream back-pressure etc work.